data/transactions/logic/eval.go

```diff
@@ -2253,33 +2298,56 @@ func (cx *evalContext) getLatestTimestamp() (timestamp uint64, err error) {
 	return uint64(ts), nil
 }
 
-func (cx *evalContext) getApplicationID() (rnd uint64, err error) {
+func (cx *EvalContext) getApplicationID() (uint64, error) {
 	if cx.Ledger == nil {
-		err = fmt.Errorf("ledger not available")
-		return
+		return 0, fmt.Errorf("ledger not available")
 	}
 	return uint64(cx.Ledger.ApplicationID()), nil
 }
 
-func (cx *evalContext) getCreatableID(groupIndex int) (cid uint64, err error) {
+func (cx *EvalContext) getApplicationAddress() (basics.Address, error) {
+	if cx.Ledger == nil {
+		return basics.Address{}, fmt.Errorf("ledger not available")
+	}
+
+	// Initialize appAddrCache if necessary
+	if cx.appAddrCache == nil {
+		cx.appAddrCache = make(map[basics.AppIndex]basics.Address)
+	}
+
+	appID := cx.Ledger.ApplicationID()
+	// Hashes are expensive, so we cache computed app addrs
+	appAddr, ok := cx.appAddrCache[appID]
+	if !ok {
+		appAddr = appID.Address()
+		cx.appAddrCache[appID] = appAddr
+	}
+
+	return appAddr, nil
+}
+
+func (cx *EvalContext) getCreatableID(groupIndex int) (cid uint64, err error) {
 	if cx.Ledger == nil {
 		err = fmt.Errorf("ledger not available")
 		return
 	}
 	return uint64(cx.Ledger.GetCreatableID(groupIndex)), nil
 }
 
-func (cx *evalContext) getCreatorAddress() ([]byte, error) {
+func (cx *EvalContext) getCreatorAddress() ([]byte, error) {
 	if cx.Ledger == nil {
 		return nil, fmt.Errorf("ledger not available")
 	}
-	addr := cx.Ledger.CreatorAddress()
```

Comment:
I agree I should remove the CreatorAddress() methods if we keep this change.  I think we should keep it, as I prefer to trim the LedgerForLogic interface, and special purpose code, when we have a more general purpose mechanism. Surely we cache things such that this is never a db lookup, since it's getting params for the currently running app.  Or don't we cache anything in roundCowState? (or perhaps below it?)